File: 240531/정수 사각형 최대 합/maximum-sum-path-in-square.java
import java.util.*;

public class Main {
    static Scanner sc = new Scanner(System.in);
    static int n = sc.nextInt();
    static int[][] dp = new int[n][n];

    public static void main(String[] args) {
        for(int r = 0; r < n; r++) {
            for(int c = 0; c < n; c++) {
                dp[r][c] = sc.nextInt();
            }
        }
        
        for(int r = 0; r < n; r++) {
            for(int c = 0; c < n; c++){
                if(c == 0 & r == 0) {
                    
                }else if(c == 0) {
                    dp[r][c] = dp[r][c] + dp[r-1][c];
                }else if(r == 0) {
                    dp[r][c] = dp[r][c] + dp[r][c-1];
                }else {
                    dp[r][c] = Math.max(dp[r][c] + dp[r][c-1], dp[r][c] + dp[r-1][c]);
                }
            }
        }        
        System.out.println(dp[n-1][n-1]);
    }
}

File: 240531/정수 사각형 최대 합/maximum-sum-path-in-square.kt
import java.util.*
import kotlin.math.*

val sc = Scanner(System.`in`)
val n = sc.nextInt()
val dp = Array(n) { IntArray(n) { sc.nextInt() } } // 0 ~ n-1

fun main() {
    for(r in 0 .. n-1) {
        for(c in 0 .. n-1) {
            when {
                (r == 0 && c == 0) -> { }
                (r == 0) -> { 
                    dp[r][c] = dp[r][c] + dp[r][c-1]
                }
                (c == 0) -> { 
                    dp[r][c] = dp[r][c] + dp[r-1][c]
                }
                else -> { 
                    dp[r][c] = max(
                        dp[r][c] + dp[r-1][c],
                        dp[r][c] + dp[r][c-1]
                    )
                }
            }
        }
    }
    println(dp[n-1][n-1])
}